### PR TITLE
Fix invalid commit range in blame operation

### PR DESCRIPTION
**Pull request by mtsgrd**
- fixes commit_to_branch related errors
- fixes big performance problem when force push disallowed

### CODE REVIEW

File: crates/gitbutler-core/src/virtual_branches/virtual.rs [hunk 1]
```diff
@@ -1695,6 +1778,9 @@ fn get_applied_status(
     target_sha: &git::Oid,
     mut virtual_branches: Vec<branch::Branch>,
 ) -> Result<(AppliedStatuses, Vec<diff::FileDiff>)> {
+    dbg!(project_repository
```
Comment (krlvi) [inline crates/gitbutler-core/src/virtual_branches/virtual.rs]: we dont wanna merge this

File: crates/gitbutler-core/src/virtual_branches/integration.rs [hunk 2]
```diff
@@ -85,14 +85,28 @@ pub fn get_workspace_head(
...
+        .filter_map(Result::ok)
+        .collect();
+
+    if heads.is_empty() {
+        heads = vec![target_commit]
+    }
+
+    // TODO: Why does commit only accept a slice of commits? Feels like we
+    // could make use of AsRef with the right traits.
+    let head_refs: Vec<&git2::Commit<'_>> = heads.iter().collect();
```
Comment (Byron) [inline crates/gitbutler-core/src/virtual_branches/integration.rs:101]: This clearly is a place where `git2` has a very inconvenient API, as a slice of refs nobody has naturally, ever. It's also a strange choice given that it needs parent-ids, not commit objects.
This goes away with `gitoxide`.

File: crates/gitbutler-core/src/virtual_branches/virtual.rs [hunk 3]
```diff
@@ -1687,6 +1688,84 @@ fn get_non_applied_status(
         .collect::<Result<Vec<_>>>()
 }
 
+fn compute_merge_base(
+    project_repository: &project_repository::Repository,
+    integration_commit: &git::Oid,
```
Comment (Byron) [inline crates/gitbutler-core/src/virtual_branches/virtual.rs:1693]: This type is `Copy` and thus shouldn't be referenced, just like (now) `ProjectId`.

File: crates/gitbutler-core/src/virtual_branches/integration.rs (hunk 2 above)
Comment (Byron) [inline crates/gitbutler-core/src/virtual_branches/integration.rs:101]: I think you can even write this:
```Rust
    let head_refs: Vec<&_> = heads.iter().collect();
```

**Review by Byron — COMMENTED** [review]
I took a quick look and have some general Rust notes.

Some of the code that moved (and was altered) looked like it's more complex than it needs to be, but I since it's in `virtual.rs` its time will come eventually :).

**Pull request merged**